### PR TITLE
feat: md support prerequisite

### LINKED ISSUE #3427: Add markdown support in the forum
Markdown support can be added to cirucuitverse form by using https://github.com/vmg/redcarpet and overriding [formated_content helper](https://github.com/excid3/simple_discussion/blob/9a4a9b553b785c5918fed236ab78a58ffdb84287/app/helpers/simple_discussion/forum_posts_helper.rb#L13)

### PR DESCRIPTION
**Pull request by VaibhavUpreti**
Fixes #3427

#### Describe the changes you have made in this PR -
prerequisite of the [PR](https://github.com/CircuitVerse/simple_discussion/pull/2) 

### Screenshots of the changes (If any) -
CURRENT VIEW..

Screenshot 2022-12-29 at 1 01 32 AM
-
Screenshot 2022-12-29 at 1 01 39 AM

The forum page is not formatted as we would have to `formatted_content`helper method [here](https://github.com/CircuitVerse/simple_discussion/blob/622094cf5fb7654fb9e9a275886fe1ddcb638f52/app/views/simple_discussion/forum_threads/_forum_thread.html.erb#L32) to make all the threads formatted as it is missing by default,Screenshot ref. previous comment.

Note: Please check **Allow edits from maintainers.** if you would like us to assist in the PR.

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -4,9 +4,8 @@ GIT
   specs:
     disposable_mail (0.1.7)
 
-GIT
-  remote: https://github.com/CircuitVerse/simple_discussion.git
-  revision: ab46323213d1172d1d0916e18cd8a8fa444f4a80
+PATH
+  remote: /Users/vaibhavupreti/Desktop/simple_discussion
```
Comment (tachyons) [inline Gemfile.lock:8]: Seems like you overrided gem in your local machine

File: Gemfile [hunk 2]
```diff
@@ -33,9 +33,9 @@ gem "terser"
 
 # Use CoffeeScript for .coffee assets and views
 gem "coffee-rails", "~> 5.0"
-
```
Comment (tachyons) [inline Gemfile:36]: Unrelated change

**Comment by tachyons** [comment]
You don't need to gem to support markdown. you can simply override the helper method from the app.

**Comment by VaibhavUpreti** [comment]
> You don't need to gem to support markdown. you can simply override the helper method from the app.

You mean that I dont need to gem in simple_discussion, and just override the method in helper, and install the gem here?

**Comment by tachyons** [comment]
Yes

**Comment by VaibhavUpreti** [comment]
Then can you pls reopen PR https://github.com/CircuitVerse/simple_discussion/pull/2So that I can remove the gem … otherwise the code looks good Sent from my iPhoneOn 28-Dec-2022, at 8:55 PM, Aboobacker MK ***@***.***> wrote:﻿
Yes

**Comment by tachyons** [comment]
We don't have to make any changes on the gem, adding redcarpet gem and
overriding helper should be in the circuitverse, not in the simple
discussion

**Comment by VaibhavUpreti** [comment]
@tachyons CURRENT VIEW..
Screenshot 2022-12-29 at 1 01 39 AM

Screenshot 2022-12-29 at 1 01 32 AM

The forum page is not formatted as we would have to helper method [here](https://github.com/CircuitVerse/simple_discussion/pull/2/files#diff-450b58e365fdbc60916dcdde1408ea79802eb2cfbc3b7d53a8fbfac02a6e3072) to make that formatted.

**Comment by VaibhavUpreti** [comment]
@tachyons I tried this approach earlier, tested these changes too
 but it not overriding the helper in simple_discussion. Unless we place the helper in app/helpers/simple_discussion/forum_posts_helper.rb and rename it as forum_posts_helper.rb
something like this...(Like it was in previous commit ).. 
`module SimpleDiscussion::ForumPostsHelper`


Also we would have to add `formatted_content` helper  [here](https://github.com/CircuitVerse/simple_discussion/blob/622094cf5fb7654fb9e9a275886fe1ddcb638f52/app/views/simple_discussion/forum_threads/_forum_thread.html.erb#L32) to make all the threads formatted as it is missing by default,Screenshot ref. previous comment.

**Comment by VaibhavUpreti** [comment]
Had to add rest of the helper methods too, as the helper seems to override the whole module instead of certain methods and was throwing up errors.

Also @tachyons  can you pls reopen [PR](https://github.com/CircuitVerse/simple_discussion/pull/2) would have to add `formatted_content` helper  [here](https://github.com/CircuitVerse/simple_discussion/blob/622094cf5fb7654fb9e9a275886fe1ddcb638f52/app/views/simple_discussion/forum_threads/_forum_thread.html.erb#L32) to make all the threads formatted as it is missing by default,Screenshot ref. previous comment.

**Comment by Tanmoy741127** [comment]
You can override the method by creating an **forum_helper.rb** in helpers folder and override the method by using alias. 
```ruby
module ForumHelper
  include SimpleDiscussion::ForumPostsHelper

  def new_formatted_content(text)
    options = %I[hard_wrap filter_html autolink tables]
    md = Redcarpet.new(text, *options).to_html
    sanitize(md)
  end

  alias formatted_content new_formatted_content
end
```

I think @tachyons was telling to do this .
@VaibhavUpreti try this changes.

**Comment by VaibhavUpreti** [comment]
Yea, you are right. This probably seems the best way to do this.

**Comment by VaibhavUpreti** [comment]
Works done on this one, just need to add formatted_content helper method [here](https://github.com/CircuitVerse/simple_discussion/blob/622094cf5fb7654fb9e9a275886fe1ddcb638f52/app/views/simple_discussion/forum_threads/_forum_thread.html.erb#L32) to make all the threads formatted as it is missing by default.

**Comment by Tanmoy741127** [comment]
> Works almost done on this one, just need to add formatted_content helper method [here](https://github.com/CircuitVerse/simple_discussion/blob/622094cf5fb7654fb9e9a275886fe1ddcb638f52/app/views/simple_discussion/forum_threads/_forum_thread.html.erb#L32) to make all the threads formatted as it is missing by default.

In the thread , whole post not shown , an truncated part only shown there. So I think we need to check , if we truncate the post body [with markdown enabled] how much that can affect the formatting.

**Comment by VaibhavUpreti** [comment]
> > Works almost done on this one, just need to add formatted_content helper method [here](https://github.com/CircuitVerse/simple_discussion/blob/622094cf5fb7654fb9e9a275886fe1ddcb638f52/app/views/simple_discussion/forum_threads/_forum_thread.html.erb#L32) to make all the threads formatted as it is missing by default.
> 
> In the thread list, whole post not shown , an truncated part only shown there. So I think we need to check , if we truncate the post body [with markdown enabled] how much that can affect the formatting.

Screenshot 2023-01-05 at 12 12 20 PM
It formats the content fairly well, but the h1's do tend to be bigger than the post name itself

**Comment by Tanmoy741127** [comment]
> > > Works almost done on this one, just need to add formatted_content helper method [here](https://github.com/CircuitVerse/simple_discussion/blob/622094cf5fb7654fb9e9a275886fe1ddcb638f52/app/views/simple_discussion/forum_threads/_forum_thread.html.erb#L32) to make all the threads formatted as it is missing by default.
> > 
> > 
> > In the thread list, whole post not shown , an truncated part only shown there. So I think we need to check , if we truncate the post body [with markdown enabled] how much that can affect the formatting.
> 
> Screenshot 2023-01-05 at 12 12 20 PM It formats the content fairly well, but the h1's do tend to be bigger than the post name itself

yes that seems may be an problem. Because as the space taken by the post grows, users need to scroll more to find an thread.

@tachyons @vedant-jain03 can suggest the correct outcome.

**Comment by tachyons** [comment]
We can skip that for now. This looks fine. We will have better markdown support with a builtin editor and preview later. something like Github and gitlab does

File: Gemfile.lock [hunk 3]
```diff
@@ -17,71 +17,71 @@ GIT
...
       mail (>= 2.7.1)
       net-imap
       net-pop
       net-smtp
-    actionmailer (7.0.3.1)
-      actionpack (= 7.0.3.1)
-      actionview (= 7.0.3.1)
-      activejob (= 7.0.3.1)
-      activesupport (= 7.0.3.1)
+    actionmailer (7.0.4)
```
Comment (tachyons) [inline Gemfile.lock:35]: This is unrelated

**Comment by tachyons** [comment]
Lint job is failing because you are using old and vulnerable version of redcarpet

**Comment by VaibhavUpreti** [comment]
@tachyons this one's good to go 👍 , though need to add `formatted_content` method [here](https://github.com/CircuitVerse/simple_discussion/blob/622094cf5fb7654fb9e9a275886fe1ddcb638f52/app/views/simple_discussion/forum_threads/_forum_thread.html.erb#L32), and then revise the commit in Gemfile.lock for simple_discussion. (to make all the threads formatted which is missing by default)

**Review by tachyons — APPROVED** [review]
